### ISSUE TITLE
Move download cache inside user defined Steam Library location

### ISSUE THREAD

**Issue by Ghostbird** [ValveSoftware/steam-for-linux#6157]
#### Your system information

* Steam client version (build number or date): Mar 20 2019 at 18:38:53, API v018, package versions 1553121671
* Distribution (e.g. Ubuntu): Debian 10 “Buster”
* Opted into Steam client beta?: Yes
* Have you checked for system updates?: Yes

#### Please describe your issue in as much detail as possible:
I have a Linux system set-up where I have a relatively small solid state drive partitioned into / and /home and a 1TB hard-disk drive where I store my data. My Steam Library is located on the hard-disk.

Today I found out that programs were crashing because the SSD was full. However, there was no real reason for it to be full. I then found out that the current Steam Client will store games in ~/.steam/steam/steamapps/downloading [i]while downloading[/i].

This is a problem for me for two reasons:

- Some games are simply too big, they would require that I keep about 30% of my entire SSD free, just to install them.
- It is highly inefficient in cases such as mine where the data is on separate file-systems, since the move from the download location to the actual final storage location is not a simple rename of the inode, but a full cross-file-system copy and erase.
Note: It is even worse when the data is on another file-system [i]on the same device[/i] such as common on Windows PCs with a single storage device divided into both C and D partitions. In this case, the reads and writes compete for device utilisation decreasing the total throughput even further.

Could you please consider moving the temporary download folder into the user designated Steam Library folder? Please consider that when a user has explicitly set a non-default Steam Library folder, there is usually a very good reason.I have a Linux system set-up where I have a relatively small solid state drive partitioned into / and /home and a 1TB hard-disk drive where I store my data. My Steam Library is located on the hard-disk.

Today I found out that programs were crashing because the SSD was full. However, there was no real reason for it to be full. I then found out that the current Steam Client will store games in ~/.steam/steam/steamapps/downloading [i]while downloading[/i].

This is a problem for me for two reasons:

- Some games are simply too big, they would require that I keep about 30% of my entire SSD free, just to install them.
- It is highly inefficient in cases such as mine where the data is on separate file-systems, since the move from the download location to the actual final storage location is not a simple rename of the inode, but a full cross-file-system copy and erase.
Note: It is even worse when the data is on another file-system [i]on the same device[/i] such as common on Windows PCs with a single storage device divided into both C and D partitions. In this case, the reads and writes compete for device utilisation decreasing the total throughput even further.

Could you please consider moving the temporary download folder into the user designated Steam Library folder? Please consider that when a user has explicitly set a non-default Steam Library folder, there is usually a very good reason.

#### Steps for reproducing this issue:

1. Install the steam client on a system with a small home partition
2. Set the Steam Library folder to a different (bigger) partition
3. Attempt to download a game that exceeds the total remaining free space on the home partition, but is smaller than the total remaining free space on the partition that holds the Steam Library.

**Comment by TehFreek** [comment]
Note that a bind mount won't necessarily solve this issue since the Steam client doesn't seem to like moving very large files (>2GB) across filesystem boundaries for some reason.

**Comment by byte1024** [comment]
I agree it should be where it will be installed, but not like that can't run out of space either so i don't necessarily see this solving a free space issue.

Maybe 1st drive where it will be installed except when low on space and in that case check home drive/partition free space for install, and if not, then fail.

**Comment by Ghostbird** [comment]
> I agree it should be where it will be installed, but not like that can't run out of space either so i don't necessarily see this solving a free space issue. I see that as a user issue (all drives can be low on free space potentially so this cannot be 100% solved except by user).
> 
> Maybe 1st check drive where it will be installed except when low on space and in that case check home drive/partition free space, and if not then fail.

I'm glad that you agree, with the location change. I agree that running out of space is a user issue. My point for this issue is not about running out of space at all. That is just an unfortunate side effect. In fact, the slowdown when moving files across filesystem boundaries is a far bigger issue, but is far more obscure, and I only managed to pinpoint it after I found this issue due to running out of disk space.

I propose that we hand the user full control and full responsibility.

Your proposal to check multiple drives seems overkill to me. The user has confirmed the library folder to use when he initialised the game installation. Steam will at that point calculate whether enough disk space is free for the installed game. The download is AFAIK always smaller than the installed game. Hence the user knows¹ that the system will not run out of space, if the download is cached on the same volume.

1) Of course some edge cases might still apply, but those are outside the control of the user and likely unknown to them. These potential issues can be considered as they arise.

**Comment by mnn** [comment]
Hit the issue just an hour ago. I have a smaller SSD for OS and `/home`. My Steam library is on a larger SSD. I had 30GB free, the update was reported by Steam to have 5GB. Steam very quickly (a minute?) allocated 15GB and most likely hit the root reserved space. Whole desktop was extra slow (reaction times 30s+), it was an agony just to kill Steam. I'll try to use a link or bind mount, but it would be better if Steam allowed the user to select the location of the download cache.

**Comment by winderine** [comment]
I ran into this issue when trying to test the operation of steam on Linux without commitment. I persistent-installed Pop OS on a thumb drive, installed steam via terminal in order to get it to access and save to my built-in SSD, and noticed that the download speed was comically slow compared to the Windows boot. Then got a message that my operating system drive was running out of space. 

As far as I can tell, this issue effectively prevents using Steam on live boot thumb drives or any situation where your operating system is on a drive smaller than the game you're downloading, which when many modern games are upwards of 50GB isn't that uncommon (Infinity Nikki, Once Human, Call of Duty Modern Warfare 3). What about OSes that run entirely in RAM? I only have 32 GB of RAM, wouldn't this make it impossible to download even a game like Palia?
